backend-tests/tests/test_deployments.py

```diff
@@ -1446,3 +1450,589 @@ class TestDeploymentsStatusUpdateEnterprise(TestDeploymentsStatusUpdateBase):
+
+        # all devices finish - and the deployment actually becomes 'finished'
+        for d in devs:
+            set_status(dep["id"], "success", d.token)
+
+        dep = get_deployment(dep["id"], user.utoken)
+        assert dep["status"] == "finished"
+
+        stats = get_stats(dep["id"], user.utoken)
+        verify_stats(stats, {"success": 10})
+
+        # an extra dev won't get this deployment
+        extra_dev = make_device_with_inventory(
+            [{"name": "foo", "value": "foo"}], user.utoken, setup_tenant.tenant_token
+        )
+        assert_get_next(204, extra_dev.token, "foo")
+
+        dep = get_deployment(dep["id"], user.utoken)
+        assert dep["status"] == "finished"
+
+        stats = get_stats(dep["id"], user.utoken)
+        verify_stats(stats, {"success": 10})
+
+    def test_deployment_ordering(self, mongo):
+        """ Check that devices only get dynamic deployments fresher than the 
+            latest one it finished.
+
+            In other words, after updating its attributes the device won't accidentally
+            fall into a deployment previous to what it tried already.
+        """
+
+        user = setup_tenant.users[0]
+
+        depfoo1 = create_dynamic_deployment(
+            "foo1", [predicate("foo", "inventory", "$eq", "foo")], user.utoken
+        )
+        depfoo2 = create_dynamic_deployment(
+            "foo2", [predicate("foo", "inventory", "$eq", "foo")], user.utoken
+        )
+        depbar = create_dynamic_deployment(
+            "bar", [predicate("foo", "inventory", "$eq", "bar")], user.utoken
+        )
+
+        # the device will ignore the 'foo' deployments, because of its inventory
+        dev = make_device_with_inventory(
+            [{"name": "foo", "value": "bar"}], user.utoken, setup_tenant.tenant_token
+        )
+        assert_get_next(200, dev.token, "bar")
+
+        # after finishing 'bar' - no other deployments qualify
+        set_status(depbar["id"], "success", dev.token)
+        assert_get_next(204, dev.token)
+
+        # after updating inventory, the device would qualify for both 'foo' deployments, but
+        # the ordering mechanism will prevent it
+        submit_inventory([{"name": "foo", "value": "foo"}], dev.token)
+        assert_get_next(204, dev.token)
+
+        # it will however get a brand new 'foo3' deployment, because it's fresher than the finished 'bar'
+        depfoo3 = create_dynamic_deployment(
+            "foo3", [predicate("foo", "inventory", "$eq", "foo")], user.utoken
+        )
+        depfoo4 = create_dynamic_deployment(
+            "foo4", [predicate("foo", "inventory", "$eq", "foo")], user.utoken
+        )
+        assert_get_next(200, dev.token, "foo3")
+
+    def test_phased_rollout(self, mongo):
+        """ Check phased rollouts with and without max_devices.
+        """
+
+        cases = [
+            # without max_devices
+            {
+                "name": "with max_devices",
```

Comment:
this one is `"without max_devices"`, right?

Reply:
yup, thanks